File: app/storage.py
from typing import Dict, List, Optional
from datetime import datetime
import uuid

class InMemoryStorage:
    def __init__(self):
        self.rooms: Dict[str, dict] = {}
        self.messages: Dict[str, List[dict]] = {}
        self.users: Dict[str, dict] = {}
        self.file_shares: Dict[str, List[dict]] = {}
    
    def create_room(self, room_id: str, room_data: dict) -> dict:
        self.rooms[room_id] = room_data
        self.messages[room_id] = []
        self.file_shares[room_id] = []
        return room_data
    
    def get_room(self, room_id: str) -> Optional[dict]:
        return self.rooms.get(room_id)
    
    def update_room_passphrase(self, room_id: str, passphrase_hash: str):
        if room_id in self.rooms:
            self.rooms[room_id]['passphrase_hash'] = passphrase_hash
            self.messages[room_id] = []
    
    def delete_room(self, room_id: str):
        self.rooms.pop(room_id, None)
        self.messages.pop(room_id, None)
        self.file_shares.pop(room_id, None)
        users_to_remove = [uid for uid, u in self.users.items() if u.get('room_id') == room_id]
        for uid in users_to_remove:
            del self.users[uid]
    
    def add_message(self, room_id: str, message: dict) -> dict:
        if room_id not in self.messages:
            self.messages[room_id] = []
        self.messages[room_id].append(message)
        return message
    
    def get_messages(self, room_id: str) -> List[dict]:
        return self.messages.get(room_id, [])
    
    def delete_message(self, room_id: str, message_id: str):
        if room_id in self.messages:
            self.messages[room_id] = [m for m in self.messages[room_id] if m['id'] != message_id]
    
    def add_user(self, user_id: str, user_data: dict) -> dict:
        self.users[user_id] = user_data
        return user_data
    
    def get_user(self, user_id: str) -> Optional[dict]:
        return self.users.get(user_id)
    
    def get_users_by_room(self, room_id: str) -> List[dict]:
        return [u for u in self.users.values() if u.get('room_id') == room_id]
    
    def remove_user(self, user_id: str):
        self.users.pop(user_id, None)
    
    def add_file_share(self, room_id: str, file_data: dict) -> dict:
        if room_id not in self.file_shares:
            self.file_shares[room_id] = []
        self.file_shares[room_id].append(file_data)
        return file_data
    
    def get_file_shares(self, room_id: str) -> List[dict]:
        return self.file_shares.get(room_id, [])

memory_storage = InMemoryStorage()

File: server/storage.ts
import { randomUUID } from "crypto";
import bcrypt from "bcryptjs";

interface ChatRoom {
  id: string;
  name: string;
  passphraseHash: string;
  createdBy: string;
  createdAt: number;
}

interface User {
  id: string;
  username: string;
  roomId: string;
  isAdmin: boolean;
  joinedAt: number;
}

interface Message {
  id: string;
  roomId: string;
  userId: string;
  username: string;
  content: string;
  timestamp: number;
  isSystem?: boolean;
}

export interface IStorage {
  createRoom(name: string, passphrase: string, createdBy: string): Promise<ChatRoom>;
  getRoom(roomId: string): Promise<ChatRoom | undefined>;
  verifyPassphrase(roomId: string, passphrase: string): Promise<boolean>;
  updateRoomPassphrase(roomId: string, newPassphrase: string): Promise<void>;
  deleteRoom(roomId: string): Promise<void>;

  addUser(user: Omit<User, "joinedAt">): Promise<User>;
  getUser(userId: string): Promise<User | undefined>;
  getUsersByRoom(roomId: string): Promise<User[]>;
  removeUser(userId: string): Promise<void>;
  getUsernameInRoom(roomId: string, username: string): Promise<User | undefined>;

  addMessage(message: Omit<Message, "id" | "timestamp">): Promise<Message>;
  getMessagesByRoom(roomId: string): Promise<Message[]>;
}

export class MemStorage implements IStorage {
  private rooms: Map<string, ChatRoom>;
  private users: Map<string, User>;
  private messages: Map<string, Message>;

  constructor() {
    this.rooms = new Map();
    this.users = new Map();
    this.messages = new Map();
  }

  async createRoom(
    name: string,
    passphrase: string,
    createdBy: string,
  ): Promise<ChatRoom> {
    const id = randomUUID();
    const passphraseHash = await bcrypt.hash(passphrase, 10);

    const room: ChatRoom = {
      id,
      name,
      passphraseHash,
      createdBy,
      createdAt: Date.now(),
    };

    this.rooms.set(id, room);
    return room;
  }

  async getRoom(roomId: string): Promise<ChatRoom | undefined> {
    return this.rooms.get(roomId);
  }

  async verifyPassphrase(roomId: string, passphrase: string): Promise<boolean> {
    const room = this.rooms.get(roomId);
    if (!room) return false;

    return bcrypt.compare(passphrase, room.passphraseHash);
  }

  async updateRoomPassphrase(
    roomId: string,
    newPassphrase: string,
  ): Promise<void> {
    const room = this.rooms.get(roomId);
    if (!room) return;

    const passphraseHash = await bcrypt.hash(newPassphrase, 10);
    room.passphraseHash = passphraseHash;
    this.rooms.set(roomId, room);

    const messagesToClear = Array.from(this.messages.values()).filter(
      (m) => m.roomId === roomId,
    );
    messagesToClear.forEach((m) => this.messages.delete(m.id));
  }

  async deleteRoom(roomId: string): Promise<void> {
    this.rooms.delete(roomId);
    
    const usersToRemove = Array.from(this.users.values()).filter(
      (u) => u.roomId === roomId,
    );
    usersToRemove.forEach((u) => this.users.delete(u.id));

    const messagesToRemove = Array.from(this.messages.values()).filter(
      (m) => m.roomId === roomId,
    );
    messagesToRemove.forEach((m) => this.messages.delete(m.id));
  }

  async addUser(user: Omit<User, "joinedAt">): Promise<User> {
    const fullUser: User = {
      ...user,
      joinedAt: Date.now(),
    };
    this.users.set(user.id, fullUser);
    return fullUser;
  }

  async getUser(userId: string): Promise<User | undefined> {
    return this.users.get(userId);
  }

  async getUsersByRoom(roomId: string): Promise<User[]> {
    return Array.from(this.users.values()).filter((u) => u.roomId === roomId);
  }

  async removeUser(userId: string): Promise<void> {
    this.users.delete(userId);
  }

  async getUsernameInRoom(
    roomId: string,
    username: string,
  ): Promise<User | undefined> {
    return Array.from(this.users.values()).find(
      (u) => u.roomId === roomId && u.username === username,
    );
  }

  async addMessage(
    message: Omit<Message, "id" | "timestamp">,
  ): Promise<Message> {
    const id = randomUUID();
    const fullMessage: Message = {
      ...message,
      id,
      timestamp: Date.now(),
    };

    this.messages.set(id, fullMessage);
    return fullMessage;
  }

  async getMessagesByRoom(roomId: string): Promise<Message[]> {
    return Array.from(this.messages.values())
      .filter((m) => m.roomId === roomId)
      .sort((a, b) => a.timestamp - b.timestamp);
  }
}

export const storage = new MemStorage();
